wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceService.java

```diff
@@ -952,11 +963,87 @@ private WorkspaceDto asDtoWithLinksAndToken(WorkspaceImpl workspace) throws Serv
       } catch (MachineTokenException e) {
         throw new ServerException(e.getMessage(), e);
       }
+
+      fixRuntimeCommandsPreviewUrls(workspaceDto);
```

Comment:
I believe this logic should not be performed on WorkspaceService layer.
Performing it once at workspace start seems right way for me https://github.com/eclipse/che/pull/14713/files